Fix package aliases to follow golang convention

feiskyer opened the pull request:
Some package aliases are not not align with golang convention https://blog.golang.org/package-names. This PR fixes them. Also adds a verify script and presubmit checks.

Fixes #35070.

cc/ @timstclair @Random-Liu

<!-- Reviewable:start -->

---

This change is [<img src="https://reviewable.kubernetes.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.kubernetes.io/reviews/kubernetes/kubernetes/35258)

<!-- Reviewable:end -->

timstclair commented:
Thanks! LGTM.

/cc @kubernetes/sig-cluster-federation 

> By the way, it's better to add a presubmit check for this, but I didn't find a easy tool. Any ideas?

I think a simple grep check should be sufficient:
```
git grep -E $'^(import |\t)[a-z]+[A-Z_][a-zA-Z]* "[^"]+"$' -- '**/*.go' ':(exclude)vendor/*' ':(exclude)staging/*'
```

feiskyer commented:
@timstclair Added presubmit hooks and rebased. PTAL.

yujuhong commented:
This is going to cause a lot of churn for other PRs. I am not sure now is the best time.

Random-Liu commented:
> This is going to cause a lot of churn for other PRs. I am not sure now is the best time.

Agree.

timstclair commented:
When would be a better time? After code freeze, but before the branch is cut? Or after master is re opened? Neither seem like great times...

We could change the precommit check to only look at additions, rather than all files, which would prevent new uses of bad aliases. Either way, the check should probably be a submit requirement (e.g. `hack/verify-pkg-names.sh`). I don't know how many developers use the pre-commit hook.

yujuhong commented:
> When would be a better time? After code freeze, but before the branch is cut? Or after master is re opened? Neither seem like great times...

I like cleanup in general, but I don't think this is urgent. There may not be a perfect time, but IMO, the beginning of a release cycle is better than two weeks before code freeze. That means after cutting the release branch, or even after the the release is out the door. This is just my two cents though.

feiskyer commented:
@timstclair @yujuhong Agree it's not a good time now, I think 'after code freeze' is a good time for this.

feiskyer commented:
Rebased. @timstclair do you think it's time for this now?

timstclair commented on hack/verify-pkg-names.sh:
```diff
@@ -0,0 +1,36 @@
+kube::golang::verify_go_version
+
+cd "${KUBE_ROOT}"
+! files=$(git grep -E $'^(import |\t)[a-z]+[A-Z_][a-zA-Z]* "[^"]+"$' -- '**/*.go' ':(exclude)vendor/*' ':(exclude)staging/*')
```
Won't errexit terminate the script at this line if packages break the convention?

An alternative is to just inline it:
```
if git grep -E $'^(import |\t)[a-z]+[A-Z_][a-zA-Z]* "[^"]+"$' -- '**/*.go' ':(exclude)vendor/*' ':(exclude)staging/*'; then
...
```

feiskyer replied:
Yes, it exits in the following if-block. Will change to the alternative way.

timstclair commented on hooks/pre-commit:
```diff
@@ -62,6 +62,20 @@ else
 echo "${reset}"
 
+echo -ne "Checking for package aliases... "
+files_break_alias=($(git grep -E $'^(import |\t)[a-z]+[A-Z_][a-zA-Z]* "[^"]+"$' -- '**/*.go' ':(exclude)vendor/*' ':(exclude)staging/*'
```
Just call `hack/verify-pkg-names.sh` from here instead

feiskyer replied:
ack

timstclair commented:
I think now is as good a time as any, as long as there's general approval for this convention. This will cause some rebase churn, but hopefully not too much.

/cc @lavalamp (buildcop)

feiskyer committed:
Add hack/verify-pkg-names.sh

Some package aliases are not not align with golang convention
https://blog.golang.org/package-names.

Fixes #35070

feiskyer committed:
Add pre-commit hooks for verifying pkg names

feiskyer commented:
@timstclair @lavalamp Addressed comments. PTAL.

timstclair commented:
LGTM. Good luck with the rebases!

feiskyer committed:
Fix package aliases to follow golang convention

feiskyer commented:
> Good luck with the rebases!

Yeah, too much rebases on this. @timstclair could you set a higher priority for this.

timstclair commented:
Bumping to p2 to avoid rebase-hell
